index.bs

```diff
@@ -679,11 +680,91 @@ are traditionally added to {{WindowEventHandlers}} rather than {{Window}}.</p>
 
 Try to design DOM events to deliver after-the-fact notifications of changes. It may be tempting to try to trigger side-effects from the action of {{EventTarget/dispatchEvent()}}, but in general this is <a href="https://lists.w3.org/Archives/Public/public-webapps/2014AprJun/0510.html">strongly discouraged</a> as it requires changes to the DOM specification when added. Your design will proceed more quickly if you avoid this pattern.
 
-<h3 id="favor-asynchronous-events">Favor asynchronous events</h3>
+<h3 id="guard-against-recursion">Guard against potential recursion</h3>
 
-A few events in the platform are specified to dispatch synchronously. These events cause problems for engines and performance issues in applications due to the possibility for re-entrant behavior they open up. The deprecated [Mutation Events](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Mutation_events), for instance, has caused many years of security issues. A more modern approach embodied in [Mutation Observers](ihttps://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) addresses most of the same use-cases in a higher-performance way which is easier to develop with and implement.
+When designing a long-running or complicated algorithm that is initiated by an 
+API call, events are appropriate for [[#events-are-for-notification|notifying]]
+user code of the ongoing process. However, they also introduce the possibility of unexpected
+re-execution of the current algorithm before it has finished! Because user code gets
+to run in the middle of the algorithm, recursion can happen if that user code triggers
+the algorithm again (directly or indirectly).
 
-If you feel you need a synchronous event in your design, please reconsider and ask the TAG for help in redesigning the API.
+If the algorithmic complexity and call graph is reasonably understood, add appropriate
+state to track that the algorithm is in progress and terminate immediately if the algorithm
+is already concurrently running. This technique is "guarding" the algorithm:
+
+<div class="example">
+    A technique (e.g., as used in the [=AbortSignal/add|AbortSignal's add algorithm=]
+    or as a stack-based variation in the [=focus update steps=])
+    for guarding against unplanned recursion. 
+    
+    In this example, an object of type <dfn noexport interface>MyObject</dfn> has
+    a method <dfn method for="MyObject">doComplexOpWithEvents()</dfn>. {{MyObject}}
+    also has an internal state attribute
+    {{MyObject}}.[[<dfn noexport attribute for="MyObject">started</dfn>]] whose initial
+    value is `false`.
+
+    The {{doComplexOpWithEvents()}} method MUST act as follows:
+    
+    1. If {{MyObject}}.[[{{started}}]] is not `false` then throw an {{InvalidStateError}}
```

Comment:
```suggestion
    1. If this.[[{{started}}]] is not `false` then throw an {{InvalidStateError}}
```

MyObject.[[started]] makes it a static property shared by the constructor; I think you want the instance instead.

Same throughout the rest of the algorithm.

Reply:
Yep, will update.